C++ API parity: ReLU6

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27309 C++ API parity: MultiheadAttention
* #27382 C++ API parity: Linear
* #27538 C++ API parity: Threshold
* #27537 C++ API parity: Tanhshrink
* #27536 C++ API parity: Tanh
* #27535 C++ API parity: Softsign
* #27534 C++ API parity: Softshrink
* #27489 C++ API parity: Softplus
* #27488 C++ API parity: Sigmoid
* #27487 C++ API parity: CELU
* #27437 C++ API parity: RReLU
* **#27436 C++ API parity: ReLU6**

Differential Revision: [D17835414](https://our.internmc.facebook.com/intern/diff/D17835414)

pbelevich committed:
C++ API parity: ReLU6

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: ReLU6"




[ghstack-poisoned]

yf225 commented on test/cpp/api/modules.cpp:
```diff
@@ -1422,3 +1439,9 @@ TEST_F(ModulesTest, PrettyPrintReLU) {
+  ASSERT_EQ(c10::str(ReLU6()), "torch::nn::ReLU6()");
+  ASSERT_EQ(c10::str(ReLU6(ReLU6Options().inplace(true))),
+    "torch::nn::ReLU6(inplace=true)");
+}
```
nit: need new line here.

yf225 commented on torch/csrc/api/src/nn/modules/activation.cpp:
```diff
@@ -134,5 +134,23 @@ void ReLUImpl::pretty_print(std::ostream& stream) const {
+void ReLU6Impl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::ReLU6(";
+  if (options.inplace()) {
+    stream << std::boolalpha  << "inplace=" << options.inplace();
```
nit:
```suggestion
    stream << std::boolalpha << "inplace=" << options.inplace();
```

yf225 commented on torch/csrc/api/include/torch/nn/modules/activation.h:
```diff
@@ -169,5 +169,28 @@ class TORCH_API ReLUImpl : public torch::nn::Cloneable<ReLUImpl> {
+class TORCH_API ReLU6Impl : public torch::nn::Cloneable<ReLU6Impl> {
+ public:
+  ReLU6Impl() : ReLU6Impl(ReLU6Options()) {}
+  explicit ReLU6Impl(const ReLU6Options& options_);
```
similarly here: would be awesome to combine these two constructors.

yf225 commented on torch/csrc/api/include/torch/nn/functional/activation.h:
```diff
@@ -52,6 +52,11 @@ inline Tensor relu(Tensor& input, const ReLUOptions& options) {
   }
 }
 
+inline Tensor relu6(Tensor& input, const ReLU6Options& options) {
```
similarly here: `options` can be optional, and we should add a test for this too.

pbelevich committed:
Update on "C++ API parity: ReLU6"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: ReLU6"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: ReLU6"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: ReLU6"




[ghstack-poisoned]

yf225 commented on test/cpp/api/functional.cpp:
```diff
@@ -409,3 +409,32 @@ TEST_F(FunctionalTest, ReLUDefaultOptions) {
+    auto x = torch::linspace(-10.0, 10.0, size * size * size);
+    x.resize_({size, size, size});
+    auto y_exp = (x < 0) * 0 + ((x >= 0) * (x <= 6)) * x + (x > 6) * 6;
+    auto y = F::relu6(x, ReLU6Options().inplace(inplace));
```
similarly here: would be great to test `F::relu6(x, /*inplace=*/inplace)`.

yf225 commented on test/cpp/api/modules.cpp:
```diff
@@ -1443,3 +1460,9 @@ TEST_F(ModulesTest, PrettyPrintReLU) {
+TEST_F(ModulesTest, PrettyPrintReLU6) {
+  ASSERT_EQ(c10::str(ReLU6()), "torch::nn::ReLU6()");
+  ASSERT_EQ(c10::str(ReLU6(ReLU6Options().inplace(true))),
+    "torch::nn::ReLU6(inplace=true)");
```
similarly here: would be great to test `ReLU6(/*inplace=*/true)`.

pbelevich committed:
Update on "C++ API parity: ReLU6"


Differential Revision: [D17835414](https://our.internmc.facebook.com/intern/diff/D17835414)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: ReLU6"


Differential Revision: [D17835414](https://our.internmc.facebook.com/intern/diff/D17835414)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: ReLU6"


Differential Revision: [D17835414](https://our.internmc.facebook.com/intern/diff/D17835414)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: ReLU6"


Differential Revision: [D17835414](https://our.internmc.facebook.com/intern/diff/D17835414)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: ReLU6"


Differential Revision: [D17835414](https://our.internmc.facebook.com/intern/diff/D17835414)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: ReLU6"


Differential Revision: [D17835414](https://our.internmc.facebook.com/intern/diff/D17835414)

[ghstack-poisoned]
